sbi/inference/posteriors/vi_posterior.py

```diff
@@ -1,69 +1,481 @@
+
+        Args:
+            method: One of {0}
+
+        Returns:
+            `NeuralPosterior` for chainable calls.
+        """
+        self._vi_method = method
+        self._optimizer_base = get_VI_method(method)
+        return self
+
+    @docstring_parameter(
+        get_default_sampling_methods(),
+        "".join(
+            [
+                "\n\t\t" + get_sampling_method_parameters_doc(n).replace("\n", "\n\t\t")
+                for n in get_default_sampling_methods()
+            ]
+        ),
+    )
     def sample(
         self,
         sample_shape: Shape = torch.Size(),
-        show_progress_bars: bool = True,
-    ):
-        raise NotImplementedError
+        x: Optional[Tensor] = None,
+        method: str = "naive",
+        **kwargs,
+    ) -> Tensor:
+        """Samples from the variational posterior distribution.
+
+        Args:
+            sample_shape: Shape of samples
+            method: Sampling method, alternatively we can debias the approximation by
+                using simple and efficient sampling schemes. We support one of {0}.
+            kwargs: Hyperparameters for the sampling methods. {1}
+
+
+        Returns:
+            Samples from posterior.
+        """
+        x = self._x_else_default_x(x)
+        if self._trained_on is None or (x != self._trained_on).all():
+            warn(
+                f"The variational posterior was not fit using observation {x}.\
+                     Please train!"
+            )
+
+        self.potential_fn.set_x(x)
+        sampling_function = get_sampling_method(method)
+        num_samples = torch.Size(sample_shape).numel()
+        samples = sampling_function(num_samples, self.potential_fn, self.q, **kwargs)
+        return samples.reshape((*sample_shape, samples.shape[-1]))
+
+    def log_prob(
+        self, theta: Tensor, x: Optional[Tensor] = None, track_gradients: bool = False
+    ) -> Tensor:
+        r"""Returns the log-probability of theta under the variational posterior.
+
+        Args:
+            theta: Parameters
+            track_gradients: Whether the returned tensor supports tracking gradients.
+                This can be helpful for e.g. sensitivity analysis, but increases memory
+                consumption.
+
+        Returns:
+            `len($\theta$)`-shaped log-probability.
+        """
+        x = self._x_else_default_x(x)
+        if self._trained_on is None or (x != self._trained_on).all():
+            warn(
+                f"The variational posterior was not fit using observation {x}.\
+                     Please train!"
+            )
+        with torch.set_grad_enabled(track_gradients):
+            theta = ensure_theta_batched(torch.as_tensor(theta))
+            return self.q.log_prob(theta)
+
+    def train(
+        self,
+        x: Optional[Array] = None,
+        n_particles: int = 256,
+        learning_rate: float = 1e-3,
+        gamma: float = 0.999,
+        max_num_iters: int = 2000,
```

Comment:
Why 2000? Why not `2**31-1`

Reply:
This was typically sufficient, but we can also increase it. But for such a large number the progress bar would be ugly :D Also if the convergence check fails then it also runs forever...

Reply:
For the sake of consistency with the rest of `sbi`, I would prefer to:
a) have no progressbar (same as for `.train()` of the inference methods)
b) rely on the convergence checks.

Have you observed that the convergence checks fail and that it would run forever?

Reply:
I always had a cutoff there so I not quite sure. For the classic think I tested the convergence check should work fine, but for different thinks I am not sure. In theory depending on the problem and also on 'n_particles' the variance of the loss even at a stationary point can be large so there it may fail. We may can increase the sensitivity 'eps' with increasing number of iterations to prevent this.

Reply:
Ok if there is a realistic scenario in which it fails then we can keep it at 2000

Reply:
I would keep it for now.